### DISCUSSION TITLE
一些经常被问到的问题

### DISCUSSION THREAD

**Discussion by guofei9987** [guofei9987/blind_watermark#52] [Q&A] [answered]
## 1. Watermark can be what? （水印可以是什么格式？）

Watermark can be
1. String
2. bits
3. image

可以是以下格式
1. 图片
2. 字符串
3. 二进制

## 2. Is there GUI/CMD? （能不能做一个界面/cmd形式？）

This Project have CMD (see documents).   
This Project will **not** have GUI. If you need it, turn to your coder friend and buy him a coffee.

目前的版本支持 command 的形式。  
经过思考，本项目的定位是开源算法组件，供其他库来调用或者封装。因此 **不会** 支持界面/GUI

## 3. Consider embed messages into meta of image? （是否考虑支持把字符串打入图片文件的头部？）

No. That is unrelated to this algorithm, and is not blind watermark.

不会有这样的功能，本项目聚焦于“把信息打入到图片的频域空间内，以此获得抗攻击性”，与之无关的功能不会做。
这些功能可以用数行代码实现：
- 如何制作图种：https://www.guofei.site/2019/06/29/crypto.html#%E5%9B%BE%E7%A7%8D
- 如何把信息打入 EXIF 中：https://www.guofei.site/2019/06/29/crypto.html#EXIF

## 4. What attack this algorithm effectively? （算法面对什么攻击会无计可施？）

Camera

手机拍屏。别侥幸！手机拍屏场景的抗攻击算法其实也存在，不过与本项目无关。

**Comment by cliffcrag** [comment]
感谢提供这么好的程序。在使用中有两个问题，烦情有空的时候看一下，一直困绕着我：

-----------------------------------------
**问题1： 字符串水印解码的时候出现'utf-8' codec can't decode byte 0xa7 in position问题。**
在用文字加密，解密的时候，小图片加密可以正常，但是解密就出现下面的问题：

_File "D:\program\Anaconda3\lib\site-packages\blind_watermark-0.2.1-py3.7.egg\blind_watermark\blind_watermark.py", line 73, in extract
wm = bytes.fromhex(hex(int(byte, base=2))[2:]).decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa7 in position 13: invalid start byte`_

如果是用附例的图片 ori_img.jpg 则没有问题。 同用img方法加水印，也会出现小点的图片出错。
请问下，对图片大小，有要求吗，是否可以设置小于多少象素的图片不打水印？ 自动拷贝原图或者不操作？

--------------------------------------------
**问题2： 请问图片要大于多少象素，才可以？**

我看了：ori_img.jpg 图是 1920X1200象素，但是我有个图是1080X608,水印同样是128X128象素的。就不行了。就会有如下提示：

_File "D:\program\Anaconda3\lib\site-packages\blind_watermark-0.2.1-py3.7.egg\blind_watermark\bwm_core.py", line 30, in init_block_index
'最多可嵌入{}kb信息，多于水印的{}kb信息，溢出'.format(self.block_num / 1000, self.wm_size / 1000))
AssertionError: 最多可嵌入10.26kb信息，多于水印的16.384kb信息，溢出_

把图片1080X608放大一倍，变成2160X1216 则就可以正常打水印了。
请问下老哥，原图尺寸，和水印图尺寸，大小有什么要求吗？

=============================

**Comment by guofei9987** [comment, marked as answer]
## 6. 对被嵌入的图片有什么要求？

1. 图片不能有太多的低频区域，例如纯色背景上面一个简单的圆，就很难打入水印。
2. 满足以上条件的情况下，图片像素只要不太小就行。一般来说像素越多，抗打击性就越强。


## 7. 图片可以嵌入多少信息

图片像素数除以64，是最多嵌入的比特数。

额外：作为图片的盲水印，其信息量不是文件大小，而是像素的积
